### DISCUSSION TITLE
What is the recommended way to inject a mock service for unit testing?

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1307] [Q&A] [answered]
I am working on a service that needs to be provided by a mock class for unit testing and by a production service when running in production mode.  So I need to inject that mock class for testing cases but inject the real class for production.  My plan is to define the service in the `AddedServices` for the language, but I don't want to have to switch between the mock service and the production service by changing code.  Is there a recommended way for injecting services that is more sophisticated than just changing the module to use a different service, or should I just define something like this?:

```

// mylanguage-module.ts
export type MyAddedServices = {
 validation:  {
   MyValidator: MyLanguageValidatior,
   MyTestScopeProvider: MyLanguageTestScopeProvider,
   MyProductionScopeProvider: MyLanguageProductionScopeProvider
  }
}
```

**Comment by msujew** [comment, marked as answer]
I actually don't believe we had the need to mock something in our languages yet. So no general recommendations. However, what I would probably do is add my to-be-mocked services to the dependency injection module context. See [this](https://github.com/eclipse-langium/langium/blob/750ff44c4d4cf3edee108b2998e818525a2b9ecc/packages/langium/src/default-module.ts#L114-L127), that we use for Langium by default. That way you could decide what and how to mock on the caller side instead of needing to build multiple different modules.

**Reply by ballcoach12** [reply]
OK.  Thanks, @msujew.  That's essentially what I am doing.  Just wanted to make sure that I'm on the right track.  The production system will have an import mechanism that hits a remote API endpoint to retrieve code to load into a Scope, so we just want to define that service for now in such a way as to delegate the responsibility for code loading to a mock object and not hit the API.

**Reply by msujew** [reply]
Yeah, that definitely makes sense 👍
